### PR TITLE
Dlg Rank

### PR DESCRIPTION
**Pull request by stevenndungu**

### CODE REVIEW

**Comment by dannyparsons** [comment]
I don't see anywhere in here where you set the name and parameters for the default function?

**Comment by stevenndungu** [comment]
Just added, deleted by mistake while deleting the previous code where functions were set. Should be ok now

File: instat/dlgRank.vb [hunk 1]
```diff
@@ -31,108 +31,80 @@ Public Class dlgRank
...
+        ucrPanelMissingValues.AddRadioButton(rdoLast, Chr(34) & "TRUE" & Chr(34))
+        ucrPanelMissingValues.SetRDefault("TRUE")
+
+
+        ucrSaveRank.SetPrefix("rank")
+        ucrSaveRank.SetSaveTypeAsColumn()
+        ucrSaveRank.SetDataFrameSelector(ucrSelectorForRank.ucrAvailableDataFrames)
+        ucrSaveRank.SetAssignToIfUncheckedValue("last_rank")
+        ucrSaveRank.SetCheckBoxText("Save Rank")
+        ucrSaveRank.SetIsComboBox()
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: This doesn't work in the same way. There should not be a checkbox but a label instead. Saving the rank is not optional so no checkbox. Instead set the text of the label.
And since there is no checkbox, there should be no use of assign to if unchecked value

File: instat/dlgRank.vb [hunk 2]
```diff
@@ -31,108 +31,80 @@ Public Class dlgRank
...
+
+        'Setting Parameters for the respective radio buttons
+        ucrPanelTies.SetParameter(New RParameter("ties.method"))
+        ucrPanelTies.AddRadioButton(rdoAverage, Chr(34) & "average" & Chr(34))
+        ucrPanelTies.AddRadioButton(rdoMinimum, Chr(34) & "min" & Chr(34))
+        ucrPanelTies.AddRadioButton(rdoMaximum, Chr(34) & "max" & Chr(34))
+        ucrPanelTies.AddRadioButton(rdoFirst, Chr(34) & "first" & Chr(34))
+        ucrPanelTies.AddRadioButton(rdoRandom, Chr(34) & "random" & Chr(34))
+        ucrPanelTies.SetRDefault(Chr(34) & "average" & Chr(34))
+        ucrPanelMissingValues.SetRDefault("FALSE")
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: How can the R default be FALSE for this?

**Comment by stevenndungu** [comment]
OK fine. Ooh! I see there is a label as well on the same ucrSave control.

File: instat/dlgRank.vb (hunk 2 above)
Comment (stevenndungu) [inline instat/dlgRank.vb]: That is there wrongly because the R default have been set to average just immediately above the line(line57)

**Comment by dannyparsons** [comment]
How is this looking now?

**Comment by stevenndungu** [comment]
It running OK.  Somebody can test it.

File: instat/dlgRank.vb [hunk 3]
```diff
@@ -16,123 +16,87 @@
...
+        clsDefaultFunction.SetRCommand("rank")
+
+        'Setting Parameters and Data types allowed
         ucrReceiverRank.Selector = ucrSelectorForRank
-        ucrReceiverRank.bUseFilteredData = False
+        ucrReceiverRank.SetMeAsReceiver()
         ucrReceiverRank.SetIncludedDataTypes({"numeric"})
-        ucrBase.iHelpTopicID = 25
+        ucrReceiverRank.SetParameter(New RParameter("x"))
+        ucrReceiverRank.SetParameterIsString()
```
Comment (lilyclements) [inline instat/dlgRank.vb]: We want this to be SetParameterIsRFunction(), otherwise we have quotes around the numbers which causes just "1"'s to be produced

File: instat/dlgRank.vb [hunk 4]
```diff
@@ -16,123 +16,87 @@
...
+        ucrPanelMissingValues.AddRadioButton(rdoLast, Chr(34) & "TRUE" & Chr(34))
+        ucrPanelMissingValues.SetRDefault("TRUE")
 
-        ucrInputColName.SetPrefix("Rank")
-        ucrInputColName.SetItemsTypeAsColumns()
-        ucrInputColName.SetDefaultTypeAsColumn()
-        ucrInputColName.SetDataFrameSelector(ucrSelectorForRank.ucrAvailableDataFrames)
-        ucrInputColName.SetValidationTypeAsRVariable()
+
+        ucrSaveRank.SetPrefix("rank")
```
Comment (lilyclements) [inline instat/dlgRank.vb]: So, on this version the "ranks" increase by

rank
rank2
rank3
...

We want the first rank to be "rank1", or the second to be rank1 if the first is rank.

(I am not sure how to fix this, however, so this is intriguing! It could be a general issue perhaps? I'm using an updated branch to this to test it out)

File: instat/dlgRank.vb [hunk 5]
```diff
@@ -16,123 +16,87 @@
...
+        ucrSaveRank.SetSaveTypeAsColumn()
+        ucrSaveRank.SetDataFrameSelector(ucrSelectorForRank.ucrAvailableDataFrames)
+        ucrSaveRank.SetLabelText("Save Rank")
```
Comment (lilyclements) [inline instat/dlgRank.vb]: We want a colon on this

File: instat/dlgRank.vb (hunk 4 above)
Comment (stevenndungu) [inline instat/dlgRank.vb]: This how it appears...
image

Comment (lilyclements) [inline instat/dlgRank.vb]: Interesting! I wasn't getting that but must have done something wrong and not realised! Apologies.

**Comment by lilyclements** [comment]
@dannyparsons This is ready for review

File: instat/dlgRank.vb [hunk 6]
```diff
@@ -16,123 +16,98 @@
...
+    ' Sub that runs only the first time the dialog loads it sets default RFunction as the base function
     Private Sub SetDefaults()
-        ucrReceiverRank.Selector = ucrSelectorForRank
-        ucrReceiverRank.SetMeAsReceiver()
-        rdoKeptAsMissing.Checked = True
-        rdoAverage.Checked = True
+        Dim clsDefaultFunction As New RFunction
+
+        ucrSelectorForRank.Reset()
         ucrSelectorForRank.Reset()
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: You are running this line twice. And you are not resetting the save control.

File: instat/dlgRank.vb [hunk 7]
```diff
@@ -16,123 +16,98 @@
...
-        rdoKeptAsMissing.Checked = True
-        TestOKEnabled()
-    End Sub
 
-    Private Sub ReopenDialog()
+        'Setting default Rfunction as the base function
+        clsDefaultFunction.SetRCommand("rank")
+        clsDefaultFunction.AddParameter("ties.method", Chr(34) & "average" & Chr(34))
+        clsDefaultFunction.AddParameter("na.last", Chr(34) & "keep" & Chr(34))
+        clsDefaultFunction.SetAssignTo("rank", strTempDataframe:=ucrSelectorForRank.ucrAvailableDataFrames.cboAvailableDataFrames.Text, strTempColumn:="rank")
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: I don't think we want this as the default. The save control should set the default each time. This is why the first column produced is "rank" instead of "rank1".
I think instead of "rank" the default value should be SaveControl.GetText()

**Comment by dannyparsons** [comment]
I assume this is ready for re-review? You need to let people know when you've finished making changes again.

File: instat/dlgRank.vb [hunk 8]
```diff
@@ -82,13 +82,13 @@ Public Class dlgRank
...
         ucrSelectorForRank.Reset()
-        ucrSelectorForRank.Reset()
+        ucrSaveRank.Reset()
 
         'Setting default Rfunction as the base function
         clsDefaultFunction.SetRCommand("rank")
         clsDefaultFunction.AddParameter("ties.method", Chr(34) & "average" & Chr(34))
         clsDefaultFunction.AddParameter("na.last", Chr(34) & "keep" & Chr(34))
-        clsDefaultFunction.SetAssignTo("rank", strTempDataframe:=ucrSelectorForRank.ucrAvailableDataFrames.cboAvailableDataFrames.Text, strTempColumn:="rank")
+        clsDefaultFunction.SetAssignTo(ucrSaveRank.GetText, strTempDataframe:=ucrSelectorForRank.ucrAvailableDataFrames.cboAvailableDataFrames.Text)
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: Why did you remove the strTempColumn from the assign to?

Comment (stevenndungu) [inline instat/dlgRank.vb]: While making final edits i went through a dialogue that had similar setassign function, where that part was not included. So I thought it wise to exclude it too.

**Comment by stevenndungu** [comment]
@dannyparsons made the correction above

File: instat/dlgRank.vb (hunk 8 above)
Comment (dannyparsons) [inline instat/dlgRank.vb]: Copying something from another dialog is not a good reason to do something. You must understand what these methods do and mean. If you don't now is the time to ask.

**Comment by lilyclements** [comment]
Final things to do - Tab Order and the ucrSave name doesn't reset upon clicking Reset (see how this is done in Combine as it will be similar to that), then this is ready for Danny to review

**Comment by stevenndungu** [comment]
@lilyclements combine text dialog uses a textbox and not a combobox. I am not sure if the ucrSave should reset to rank1, because if resets to rank1 then it will imply that it has to replace existing rank1 which i doubt should happen

**Comment by lilyclements** [comment]
The text and combo box have the same options for what we want to do.
We want to press reset and for it to reset to RankX, where X is the next number (i.e., if we have Rank1, Rank2, and I press reset I want it to say Rank3). At the moment if I input something into the combobox and press reset, it doesn't reset to RankX, it just stays as the thing I inputted in

**Comment by stevenndungu** [comment]
@lilyclements combine text dialog uses a textbox and not a combobox. I see it working well, that is, resetting to the previous prefix name set if the user types different name from the prefix name given by default.  
Or may be you could explain the anomaly in the reset.

**Comment by lilyclements** [comment]
I know, but whether it's a text box or combo box is irrelevant for this issue. The issue is no longer occurring so perhaps I had it on a different branch or something!
This is ready for @dannyparsons

File: instat/dlgRank.vb [hunk 9]
```diff
@@ -16,123 +16,98 @@
...
 
     End Sub
+
     Private Sub InitialiseDialog()
-        ucrBase.clsRsyntax.SetFunction("rank")
+        ucrBase.iHelpTopicID = 25
+
+        'Setting Parameters and Data types allowed
         ucrReceiverRank.Selector = ucrSelectorForRank
-        ucrReceiverRank.bUseFilteredData = False
```
Comment (dannyparsons) [inline instat/dlgRank.vb]: Why did you remove this line?

Comment (stevenndungu) [inline instat/dlgRank.vb]: I must have deleted by mistake while deleting the previous code- on my first commit. Thanks for noting this.

**Comment by stevenndungu** [comment]
the line added

File: instat/dlgRank.vb (hunk 9 above)
Comment (dannyparsons) [inline instat/dlgRank.vb]: That's ok. Do you know why this line is needed?

Comment (stevenndungu) [inline instat/dlgRank.vb]: Yes, it determines if it's the whole data set or subset of the data (filtered data) to be used in calculations by a specific dialogue. If set to True the whole data set is always used, while if it is False then the current filtered data is used. *correct me if am wrong

Comment (dannyparsons) [inline instat/dlgRank.vb]: Yep that's correct.

**Comment by stevenndungu** [comment]
I think all changes are made on this... unless there is anything more from reviewers?

**Comment by maxwellfundi** [comment]
@dannyparsons  you could now have a look at this.
